### ISSUE TITLE
Unable to set custom gradient method in `QGAN` generator

### ISSUE THREAD

**Issue by ElePT** [Qiskit/qiskit-machine-learning#394]
### Environment

- **Qiskit Machine Learning version**: 0.4.0
- **Python version**: 3.10.0
- **Operating system**: macOS Monterrey (12.2.1)

### What is happening?

When specifying a custom gradient in the QGAN generator, a `ValueError` gets raised during training.

### How can we reproduce the issue?

On [04_qgans_for_loading_random_distributions.ipynb](https://github.com/Qiskit/qiskit-machine-learning/blob/main/docs/tutorials/04_qgans_for_loading_random_distributions.ipynb), try to set a custom gradient with either finite differences or parameter shift method, by typing the following code into the **Initialize the qGAN** section:

```
from qiskit.opflow import Gradient

custom_grad = Gradient("fin_diff")
# custom_grad = Gradient("param_shift")

# Set quantum generator
qgan.set_generator(generator_circuit=g_circuit, generator_init_params=init_params,
                   generator_gradient=custom_grad
                   )
```

This cell should run without issues, but if you continue with the tutorial, the cell with:

```
result = qgan.run(quantum_instance)
```

Will raise the following error:

```
ValueError: shapes (3,1) and (4,4) not aligned: 1 (dim 1) != 4 (dim 0)
```

### What should happen?

The user should be able to set a custom gradient method. 

### Any suggestions?

I did not have time to look into the issue, but the error traceback points to the loss calculation in [quantum_generator.py](https://github.com/Qiskit/qiskit-machine-learning/blob/main/qiskit_machine_learning/algorithms/distribution_learners/qgan/quantum_generator.py)

The exposed traceback is the following:

```
ValueError                                Traceback (most recent call last)
File ~/Desktop/qiskit-machine-learning/qiskit_machine_learning/algorithms/distribution_learners/qgan/quantum_generator.py:353, in QuantumGenerator.loss(self, x, weights)
    351 try:
    352     # pylint: disable=no-member
--> 353     loss = (-1) * np.dot(np.log(x).transpose(), weights)
    354 except Exception:  # pylint: disable=broad-except

File <__array_function__ internals>:5, in dot(*args, **kwargs)

ValueError: shapes (1,3) and (4,4) not aligned: 3 (dim 1) != 4 (dim 0)

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
Input In [120], in <cell line: 2>()
      1 # Run qGAN
----> 2 result = qgan.run(quantum_instance)

File ~/Desktop/qiskit-machine-learning/qiskit_machine_learning/algorithms/distribution_learners/qgan/qgan.py:204, in QGAN.run(self, quantum_instance, **kwargs)
    201     if quantum_instance is not None:
    202         self.quantum_instance = quantum_instance
--> 204 return self._run()

File ~/Desktop/qiskit-machine-learning/qiskit_machine_learning/algorithms/distribution_learners/qgan/qgan.py:479, in QGAN._run(self)
    473 if self._quantum_instance.backend_name == ("unitary_simulator" or "clifford_simulator"):
    474     raise QiskitMachineLearningError(
    475         "Chosen backend not supported - "
    476         "Set backend either to statevector_simulator, qasm_simulator"
    477         " or actual quantum hardware"
    478     )
--> 479 self.train()
    481 return self._ret

File ~/Desktop/qiskit-machine-learning/qiskit_machine_learning/algorithms/distribution_learners/qgan/qgan.py:427, in QGAN.train(self)
    425     # 2. Train Generator
    426     self._generator.discriminator = self._discriminator
--> 427     ret_g = self._generator.train(self._quantum_instance, shots=self._batch_size)
    428     g_loss_min = ret_g["loss"]
    430 self._d_loss.append(np.around(float(d_loss_min), 4))

File ~/Desktop/qiskit-machine-learning/qiskit_machine_learning/algorithms/distribution_learners/qgan/quantum_generator.py:476, in QuantumGenerator.train(self, quantum_instance, shots)
    471     self._gradient_function = self._convert_to_gradient_function(
    472         self._gradient_function, quantum_instance, self._discriminator
    473     )
    475 objective = self._get_objective_function(quantum_instance, self._discriminator)
--> 476 opt_result = self._optimizer.minimize(
    477     fun=objective,
    478     x0=self._bound_parameters,
    479     jac=self._gradient_function
    480 )
    481 self._bound_parameters = opt_result.x
    482 self._ret["loss"] = opt_result.fun

File /opt/anaconda3/envs/qml/lib/python3.10/site-packages/qiskit/utils/deprecation.py:27, in deprecate_arguments.<locals>.decorator.<locals>.wrapper(*args, **kwargs)
     25 if kwargs:
     26     _rename_kwargs(func.__name__, kwargs, kwarg_map)
---> 27 return func(*args, **kwargs)

File /opt/anaconda3/envs/qml/lib/python3.10/site-packages/qiskit/algorithms/optimizers/adam_amsgrad.py:236, in ADAM.minimize(self, fun, x0, jac, bounds, objective_function, initial_point, gradient_function)
    233 if jac is None:
    234     jac = Optimizer.wrap_function(Optimizer.gradient_num_diff, (fun, self._eps))
--> 236 derivative = jac(x0)
    237 self._t = 0
    238 self._m = np.zeros(np.shape(derivative))

File ~/Desktop/qiskit-machine-learning/qiskit_machine_learning/algorithms/distribution_learners/qgan/quantum_generator.py:424, in QuantumGenerator._convert_to_gradient_function.<locals>.gradient_function(current_point)
    422 value_dict = {free_params[i]: current_point[i] for i in range(len(free_params))}
    423 analytical_gradients = np.array(grad_object.assign_parameters(value_dict).eval())
--> 424 loss_gradients = self.loss(
    425     prediction_generated, np.transpose(analytical_gradients)
    426 ).real
    427 return loss_gradients

File ~/Desktop/qiskit-machine-learning/qiskit_machine_learning/algorithms/distribution_learners/qgan/quantum_generator.py:355, in QuantumGenerator.loss(self, x, weights)
    353     loss = (-1) * np.dot(np.log(x).transpose(), weights)
    354 except Exception:  # pylint: disable=broad-except
--> 355     loss = (-1) * np.dot(np.log(x), weights)
    356 return loss.flatten()

File <__array_function__ internals>:5, in dot(*args, **kwargs)

ValueError: shapes (3,1) and (4,4) not aligned: 1 (dim 1) != 4 (dim 0)

```

**Comment by ElePT** [comment]
@Zoufalc, @ebergamot, you might want to take a look into this too.

**Comment by ebermot** [comment]
Hi @ElePT, I'm currently working on it and I'd like to be assigned to this issue.

**Comment by adekusar-drl** [comment]
Closed in #405
